growthbook.go

```diff
@@ -97,136 +346,277 @@ func (fr *FeatureResult) GetValueWithDefault(def FeatureValue) FeatureValue {
+// simple.)
+func (gb *GrowthBook) Run(exp *Experiment) *Result {
+	gb.inner.RLock()
+	defer gb.inner.RUnlock()
+
+	result := gb.doRun(exp, "")
+	gb.fireSubscriptions(exp, result)
+	return result
+}
 
-		// Otherwise, return experiment result.
-		return getFeatureResult(result.Value, ExperimentResultSource, &experiment, result)
+// Subscribe adds a callback that is called every time GrowthBook.Run
+// is called. This is different from the tracking callback since it
+// also fires when a user is not included in an experiment.
+func (gb *GrowthBook) Subscribe(callback ExperimentCallback) func() {
+	gb.inner.Lock()
+	defer gb.inner.Unlock()
+
+	id := gb.inner.nextSubscriptionID
+	gb.inner.subscriptions[id] = callback
+	gb.inner.nextSubscriptionID++
+	return func() {
+		delete(gb.inner.subscriptions, id)
+	}
+}
+
+// GetAllResults returns a map containing all the latest results from
+// all experiments that have been run, indexed by the experiment key.
+func (gb *GrowthBook) GetAllResults() map[string]*Assignment {
+	gb.inner.RLock()
+	defer gb.inner.RUnlock()
+
+	return gb.inner.assigned
+}
+
+// ClearSavedResults clears out any experiment results saved within a
+// GrowthBook instance (used for deciding whether to send data to
+// subscriptions).
+func (gb *GrowthBook) ClearSavedResults() {
+	gb.inner.Lock()
+	defer gb.inner.Unlock()
+
+	gb.inner.assigned = make(map[string]*Assignment)
+}
+
+// ClearTrackingData clears out records of calls to the experiment
+// tracking callback.
+func (gb *GrowthBook) ClearTrackingData() {
+	gb.inner.Lock()
+	defer gb.inner.Unlock()
+
+	gb.inner.trackedExperiments = make(map[string]bool)
+}
+
+// GetAPIInfo gets the hostname and client key for GrowthBook API
+// access.
+func (gb *GrowthBook) GetAPIInfo() (string, string) {
+	gb.inner.RLock()
+	defer gb.inner.RUnlock()
+
+	apiHost := gb.inner.context.APIHost
+	if apiHost == "" {
+		apiHost = "https://cdn.growthbook.io"
+	}
+
+	return strings.TrimRight(apiHost, "/"), gb.inner.context.ClientKey
+}
+
+type FeatureRepoOptions struct {
+	AutoRefresh bool
+	Timeout     time.Duration
+	SkipCache   bool
+}
+
+func (gb *GrowthBook) LoadFeatures(options *FeatureRepoOptions) {
+	gb.refresh(options, true, true)
+	if options != nil && options.AutoRefresh {
+		RepoSubscribe(gb)
+	}
+}
+
+func (gb *GrowthBook) RefreshFeatures(options *FeatureRepoOptions) {
+	gb.refresh(options, false, true)
+}
+
+func (gb *GrowthBook) refresh(
+	options *FeatureRepoOptions, allowStale bool, updateInstance bool) {
+
+	if gb.inner.context.ClientKey == "" {
+		logError("Missing clientKey")
+		return
+	}
+	var timeout time.Duration
+	skipCache := gb.inner.context.DevMode
+	if options != nil {
+		timeout = options.Timeout
+		skipCache = skipCache || options.SkipCache
+	}
+	ConfigureCacheStaleTTL(gb.inner.context.CacheTTL)
+	RepoRefreshFeatures(gb, timeout, skipCache, allowStale, updateInstance)
+}
+
+//-- PRIVATE FUNCTIONS START HERE ----------------------------------------------
```

Comment:
Looks like the above `func (gb *GrowthBook) refresh(...)` is also private, do we want to move that into this section or remove the private functions indicator?

Reply:
Agreed. I moved it.